File: src/main/java/org/usfirst/irs1318/gamesim/engine/event/EventQueue.java
package org.usfirst.irs1318.gamesim.engine.event;

import java.util.Comparator;
import java.util.PriorityQueue;

public class EventQueue {
    private final PriorityQueue<Event> queue = new PriorityQueue<>(Comparator.comparing(Event::getTime));

    public void addEvent(Event event) {
        queue.offer(event);
    }

    public void removeEvent(Event event) {
        queue.remove(event);
    }

    public Event getNext() {
        return queue.poll();
    }

    public Event peekNext() {
        return queue.peek();
    }

    public PriorityQueue<Event> getQueue() {
        return queue;
    }
}

File: src/main/scala/org/usfirst/irs1318/gamesim/engine/event/EventQueue.scala
package org.usfirst.irs1318.gamesim.engine.event

import scala.collection.immutable.SortedSet

case class EventQueue(private val queue: SortedSet[Event] = SortedSet[Event]()) {
  @inline def mapQueue(fn: SortedSet[Event] => SortedSet[Event]): EventQueue = copy(queue = fn(queue))

  def +(event: Event): EventQueue = mapQueue(_ + event)

  def -(event: Event): EventQueue = mapQueue(_ - event)

  def peek(): Option[Event] = {
    if (queue.isEmpty) None else Some(queue.firstKey)
  }

  def pop(): Option[(Event, EventQueue)] = {
    if (queue.isEmpty) {
      None
    } else {
      val first = queue.firstKey
      Some((first, this - first))
    }
  }
}
